Core/src/org/sleuthkit/autopsy/threadutils/TaskRetryUtil.java

```diff
@@ -0,0 +1,280 @@
+ */
+package org.sleuthkit.autopsy.threadutils;
+
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A utility that attempts a task a specified number of times with a specified
+ * delay before each attempt and an optional timeout for each attempt. If an
+ * attempt times out, the attempt will be cancelled and the next attempt, if
+ * any, will begin.
+ */
+public class TaskRetryUtil {
+
+    private static final AtomicLong totalTasks = new AtomicLong();
+    private static final AtomicLong totalTaskRetries = new AtomicLong();
+    private static final AtomicLong totalTaskAttemptTimeOuts = new AtomicLong();
+    private static final AtomicLong totalFailedTasks = new AtomicLong();
+
+    /**
+     * Encapsulates the specification of a task attempt for the attemptTask()
+     * utility.
+     */
+    public static class TaskAttempt {
+
+        private final Long delay;
+        private final Long timeOut;
+        private final TimeUnit timeUnit;
+
+        /**
+         * Constructs an object that encapsulates the specification of a task
+         * attempt for the attemptTask() utility. The attempt will have neither
+         * a delay nor a time out.
+         *
+         * @param delay         The delay before the task should be attempted,
+         *                      may be zero or any positive integer.
+         * @param delayTimeUnit The time unit for the delay before the task
+         *                      should be attempted.
+         */
+        public TaskAttempt() {
+            this.delay = 0L;
+            this.timeOut = 0L;
+            this.timeUnit = TimeUnit.SECONDS;
+        }
+
+        /**
+         * Constructs an object that encapsulates the specification of a task
+         * attempt for the attemptTask() utility.
+         *
+         * @param delay    The delay before the task should be attempted, may be
+         *                 zero or any positive integer.
+         * @param timeUnit The time unit for the delay before the task should be
+         *                 attempted.
+         */
+        public TaskAttempt(Long delay, TimeUnit timeUnit) {
+            if (delay == null || delay < 0) {
+                throw new IllegalArgumentException(String.format("Argument for delay parameter = %d, must be zero or any positive integer", delay));
+            }
+            if (timeUnit == null) {
+                throw new IllegalArgumentException("Argument for timeUnit parameter is null");
+            }
+            this.delay = delay;
+            this.timeOut = 0L;
+            this.timeUnit = TimeUnit.SECONDS;
```

Comment:
I think this should be saving the passed in timeUnit

Reply:
Yes, good catch, thanks.